forem/forem
# Remove Database.table_available?

## What type of PR is this? (check all applicable)


## Description

PR #14138 added a call to `Database.table_available?`. This unfortunately hit the database every single time. If we're hitting the DB on every setting check, we've negated the value of caching.

In the table below, you can see that this extra DB query executes 10-12x as much as our previously most frequently executed query, putting pretty immense strain on the database. These queries are unlikely to use indexes since there are only a few dozen rows in the table at most, so they won't hit Postgres's minimum-tuple threshold to invoke an index scan, so these will _always_ be sequential scans. 

<img width="348" alt="Showing our top queries on Datadog" src="https://user-images.githubusercontent.com/108205/131542352-663d66e8-f520-41d3-b330-a5ec9bdfebeb.png">

This PR replaces `Database.table_available?` calls with `Model.table_exists?` calls. ActiveRecord already implements this at the model level and also makes use of the ActiveRecord schema cache. `Database.table_available?` did not use this query cache and ended up sending those queries to Postgres every single time.

Since `Profile` was the only other usage of this method, I swapped it over to `table_exists?`, too, and removed `Database` entirely. I'm not 100% sure of this is the right move yet, because it rescued some exceptions to return `false`. Worst-case scenario, we can instead define a method on `ApplicationRecord` so it can still use the schema cache.

## Related Tickets & Documents

- #14138 
- #14637 

## QA Instructions, Screenshots, Recordings

- In a Rails console: `1000.times { Settings::General.app_domain }`
- Check your Postgres logs (not your Rails logs, it's hidden from the ActiveRecord logger)
  - A query resembling `SELECT c.relname FROM pg_class c LEFT JOIN pg_namespace n ON n.oid = c.relnamespace WHERE n.nspname = ANY (current_schemas(false)) AND c.relname = 'site_configs' AND c.relkind IN ('r','p')` should only execute once.

## Added/updated tests?

- [x] No, and this is why: Behavior other than caching has not changed and that cache isn't something we own

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Admin Guide](https://admin.forem.com/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
      or in a [forem.dev](http://forem.dev) post
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

- [This Datadog chart](https://app.datadoghq.com/apm/service/practical_developer-postgres/postgres.query?env=production&sort=duration%2Cdesc&topGraphs=latency%3Alatency%2Cerrors%3Acount%2Chits%3Acount&start=1630426845589&end=1630430445589&paused=false) should have little to no occurrences instead of > 1k per second

## [optional] What gif best describes this PR or how it makes you feel?

![Cache is king](https://f4.bcbits.com/img/a0456218977_10.jpg)
